stellargraph/connector/neo4j/sampler.py

```diff
@@ -0,0 +1,202 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__all__ = [
+    "Neo4JSampledBreadthFirstWalk",
+    "Neo4JDirectedBreadthFirstNeighbors",
+]
+
+
+import numpy as np
+import warnings
+from collections import defaultdict, deque
+
+from ...core.schema import GraphSchema
+from ...core.graph import StellarGraph
+from ...data.explorer import GraphWalk
+from ...core.experimental import experimental
+
+
+@experimental(reason="the class is not fully tested")
+class Neo4JSampledBreadthFirstWalk(GraphWalk):
+    """
+    Breadth First Walk that generates a sampled number of paths from a starting node.
+    It can be used to extract a random sub-graph starting from a set of initial nodes from Neo4J database.
+    """
+
+    def run(self, neo4j_graphdb, nodes=None, n=1, n_size=None, seed=None):
+        """
+        Send queries to Neo4J graph databases and collect sampled breadth-first walks starting from the root nodes.
+
+        Args:
+            neo4j_graphdb: (py2neo.Graph) the Neo4J Graph Database object
+            nodes (list): A list of root node ids such that from each node n BFWs will be generated up to the
+            given depth d.
+            n_size (int): The number of neighbouring nodes to expand at each depth of the walk. Sampling of
+            n (int, default 1): Number of walks per node id.
+            neighbours with replacement is always used regardless of the node degree and number of neighbours
+            requested.
+            seed (int, optional): Random number generator seed; default is None
+
+        Returns:
+            A list of lists, each list is a sequence of sampled node ids at a certain hop.
+        """
+
+        def bfs_neighbor_query(sampling_direction):
+            direction = {"BOTH": "--", "IN": "<--", "OUT": "-->"}
```

Comment:
To make the code slightly nicer inside the string, this could store the arrow directly:

```suggestion
            direction_arrow = {"BOTH": "--", "IN": "<--", "OUT": "-->"}[sampling_direction]
            
            f"""
                ...
                MATCH (cur_node){direction_arrow}(neighbors)
                ...
            """
```